keras_nlp/models/gemma/gemma_attention.py

```diff
@@ -92,10 +92,13 @@ def build(self, inputs_shape):
     def _apply_rope(self, x, positions):
         """Rope rotate q or k."""
         # TODO: refactor to use RotaryEmbedding layer?
+        x = ops.cast(
```

Comment:
do we really want `x` in float32? that feels a little awkward, given that we are going to downcast immediately. seems reasonable to keep the `sin` and `cos` line at full precision (though i have no idea if that's important or not).

but this line seems like it should probably be done in `compute_dtype` => `ops.stack([x1 * cos - x2 * sin, x2 * cos + x1 * sin], axis=-1)`

so
```
sin = ops.cast(ops.sin(radians), self.compute_dtype)
cos = ops.cast(ops.cos(radians), self.compute_dtype)
x1, x2 = ops.split(x, 2, axis=-1)
return ops.stack([x1 * cos - x2 * sin, x2 * cos + x1 * sin], axis=-1)
```

incidentally, that would mean that if we ever did cache the rotary vectors, that would mean we are caching them at `float32` and applying them at the correct dtype for the model

wdyt?

Reply:
I agree, casting x is a bit awkward. I guess as long as radians/positions/timescale are in float32 initially we loose the numerical inacuracy. 

Dug more into the other submited fix for sin/cos in float32 https://github.com/huggingface/transformers/pull/29285#issuecomment-1966339811 - and it seems like this is the right call here (downcasting before we stack/multiply).

Reply:
Sweet! This LGTM! Will pull it in.